File: core/src/main/java/com/yuvalshavit/antlr4/DenterHelper.java
package com.yuvalshavit.antlr4;

import org.antlr.v4.runtime.CommonToken;
import org.antlr.v4.runtime.Token;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Queue;

public abstract class DenterHelper {
  private final Queue<Token> dentsBuffer = new ArrayDeque<>();
  private final Deque<Integer> indentations = new ArrayDeque<>();
  private final int nlToken;
  private final int indentToken;
  private final int dedentToken;
  private boolean reachedEof;
  private EofHandler eofHandler = new StandardEofHandler();

  protected DenterHelper(int nlToken, int indentToken, int dedentToken) {
    this.nlToken = nlToken;
    this.indentToken = indentToken;
    this.dedentToken = dedentToken;
  }

  public Token nextToken() {
    initIfFirstRun();
    Token t = dentsBuffer.isEmpty()
      ? pullToken()
      : dentsBuffer.remove();
    if (reachedEof) {
      return t;
    }
    final Token r;
    if (t.getType() == nlToken) {
      r = handleNewlineToken(t);
    } else if (t.getType() == Token.EOF) {
      r = eofHandler.apply(t);

    } else {
      r = t;
    }
    return r;
  }

  public DenterOptions getOptions() {
    return new DenterOptionsImpl();
  }

  protected abstract Token pullToken();

  private void initIfFirstRun() {
    if (indentations.isEmpty()) {
      indentations.push(0);
      // First invocation. Look for the first non-NL. Enqueue it, and possibly an indentation if that non-NL
      // token doesn't start at char 0.
      Token firstRealToken;
      do {
        firstRealToken = pullToken();
      }
      while(firstRealToken.getType() == nlToken);

      if (firstRealToken.getCharPositionInLine() > 0) {
        indentations.push(firstRealToken.getCharPositionInLine());
        dentsBuffer.add(createToken(indentToken, firstRealToken));
      }
      dentsBuffer.add(firstRealToken);
    }
  }

  private Token handleNewlineToken(Token t) {
    // fast-forward to the next non-NL
    Token nextNext = pullToken();
    while (nextNext.getType() == nlToken) {
      t = nextNext;
      nextNext = pullToken();
    }
    if (nextNext.getType() == Token.EOF) {
      return eofHandler.apply(nextNext);
    }
    // nextNext is now a non-NL token; we'll queue it up after any possible dents

    String nlText = t.getText();
    int indent = nlText.length() - 1; // every NL has one \n char, so shorten the length to account for it
    if (indent > 0 && nlText.charAt(0) == '\r') {
      --indent; // If the NL also has a \r char, we should account for that as well
    }
    int prevIndent = indentations.peek();
    final Token r;
    if (indent == prevIndent) {
      r = t; // just a newline
    } else if (indent > prevIndent) {
      r = createToken(indentToken, t);
      indentations.push(indent);
    } else {
      r = unwindTo(indent, t);
    }
    dentsBuffer.add(nextNext);
    return r;
  }

  private final class StandardEofHandler implements EofHandler {
    @Override
    public Token apply(Token t) {
      Token r;
      // when we reach EOF, unwind all indentations. If there aren't any, insert a NL. This lets the grammar treat
      // un-indented expressions as just being NL-terminated, rather than NL|EOF.
      if (indentations.isEmpty()) {
        r = createToken(nlToken, t);
        dentsBuffer.add(t);
      } else {
        r = unwindTo(0, t);
        dentsBuffer.add(t);
      }
      reachedEof = true;
      return r;
    }
  }

  private Token createToken(int tokenType, Token copyFrom) {
    String tokenTypeStr;
    if (tokenType == nlToken) {
      tokenTypeStr = "newline";
    } else if (tokenType == indentToken) {
      tokenTypeStr = "indent";
    } else if (tokenType == dedentToken) {
      tokenTypeStr = "dedent";
    } else {
      tokenTypeStr = null;
    }
    CommonToken r = new InjectedToken(copyFrom, tokenTypeStr);
    r.setType(tokenType);
    return r;
  }

  /**
   * Returns a DEDENT token, and also queues up additional DEDENTS as necessary.
   * @param targetIndent the "size" of the indentation (number of spaces) by the end
   * @param copyFrom the triggering token
   * @return a DEDENT token
   */
  private Token unwindTo(int targetIndent, Token copyFrom) {
    assert dentsBuffer.isEmpty() : dentsBuffer;
    dentsBuffer.add(createToken(nlToken, copyFrom));
    // To make things easier, we'll queue up ALL of the dedents, and then pop off the first one.
    // For example, here's how some text is analyzed:
    //
    //  Text          :  Indentation  :  Action         : Indents Deque
    //  [ baseline ]  :  0            :  nothing        : [0]
    //  [   foo    ]  :  2            :  INDENT         : [0, 2]
    //  [    bar   ]  :  3            :  INDENT         : [0, 2, 3]
    //  [ baz      ]  :  0            :  DEDENT x2      : [0]

    while (true) {
      int prevIndent = indentations.pop();
      if (prevIndent == targetIndent) {
        break;
      }
      if (targetIndent > prevIndent) {
        // "weird" condition above
        indentations.push(prevIndent); // restore previous indentation, since we've indented from it
        dentsBuffer.add(createToken(indentToken, copyFrom));
        break;
      }
      dentsBuffer.add(createToken(dedentToken, copyFrom));
    }
    indentations.push(targetIndent);
    return dentsBuffer.remove();
  }

  private interface EofHandler {
    Token apply(Token t);
  }

  private class DenterOptionsImpl implements DenterOptions {
    @Override
    public void ignoreEOF() {
      eofHandler = new EofHandler() {
        @Override
        public Token apply(Token t) {
          reachedEof = true;
          return t;
        }
      };
    }

  }

  private static class InjectedToken extends CommonToken {
    private String type;

    private InjectedToken(Token oldToken, String type) {
      super(oldToken);
      this.type = type;
    }

    @Override
    public String getText() {
      if (type != null) {
        setText(type);
      }
      return super.getText();
    }
  }

  public interface Builder0 {
    Builder1 nl(int nl);
  }

  public interface Builder1 {
    Builder2 indent(int indent);
  }

  public interface Builder2 {
    Builder3 dedent(int dedent);
  }

  public interface Builder3 {
    DenterHelper pullToken(TokenPuller puller);
  }

  public interface TokenPuller {
    Token pullToken();
  }

  public static Builder0 builder() {
    return new BuilderImpl();
  }

  private static class BuilderImpl implements Builder0, Builder1, Builder2, Builder3 {
    private int nl;
    private int indent;
    private int dedent;

    @Override
    public Builder1 nl(int nl) {
      this.nl = nl;
      return this;
    }

    @Override
    public Builder2 indent(int indent) {
      this.indent = indent;
      return this;
    }

    @Override
    public Builder3 dedent(int dedent) {
      this.dedent = dedent;
      return this;
    }

    @Override
    public DenterHelper pullToken(TokenPuller puller) {
      final TokenPuller p = puller;
      return new DenterHelper(nl, indent, dedent) {
        @Override
        protected Token pullToken() {
          return p.pullToken();
        }
      };
    }
  }
}

File: Python3/antlr_denter/DenterHelper.py
import types

from antlr4.Token import CommonToken
from antlr4.Token import Token


class DenterHelper(object):
    def __init__(self, nl_token, indent_token, dedent_token, should_ignore_eof):
        self.dents_buffer = []
        self.indentations = []
        self.nl_token = nl_token
        self.indent_token = indent_token
        self.dedent_token = dedent_token
        self.reached_eof = False
        self.should_ignore_eof = should_ignore_eof

    def next_token(self):
        self.init_if_first_run()
        if not self.dents_buffer:
            t = self.pull_token()
        else:
            t = self.dents_buffer.pop(0)
        if self.reached_eof:
            return t
        if t.type == self.nl_token:
            r = self.handle_newline_token(t)
        elif t.type == Token.EOF:
            r = self.apply(t)
        else:
            r = t
        return r

    def pull_token(self):
        """

        :rtype: CommonToken
        """
        pass

    def init_if_first_run(self):
        if not self.indentations:
            self.indentations.insert(0, 0)
            while True:
                first_real_token = self.pull_token()
                if first_real_token.type != self.nl_token:
                    break
            if first_real_token.column > 0:
                self.indentations.insert(0, first_real_token.column)
                self.dents_buffer.append(self.create_token(self.indent_token, first_real_token))
            self.dents_buffer.append(first_real_token)

    def handle_newline_token(self, t: Token):
        next_next = self.pull_token()
        while next_next.type == self.nl_token:
            t = next_next
            next_next = self.pull_token()
        if next_next.type == Token.EOF:
            return self.apply(next_next)
        nl_text = t.text
        indent = len(nl_text) - 1
        if indent > 0 and nl_text[0] == '\r':
            indent -= 1
        prev_indent = self.indentations[0]
        if indent == prev_indent:
            r = t
        elif indent > prev_indent:
            r = self.create_token(self.indent_token, t)
            self.indentations.insert(0, indent)
        else:
            r = self.unwind_to(indent, t)
        self.dents_buffer.append(next_next)
        return r

    def create_token(self, token_type, copy_from: CommonToken):
        if token_type == self.nl_token:
            token_type_str = 'newLine'
        elif token_type == self.indent_token:
            token_type_str = 'indent'
        elif token_type == self.dedent_token:
            token_type_str = 'dedent'
        else:
            token_type_str = None
        r = self.get_injected_token(copy_from, token_type_str)
        r.type = token_type
        return r

    def get_injected_token(self, copy_from: CommonToken, token_type_str):
        new_token = copy_from.clone()
        new_token.text = token_type_str
        return new_token

    def unwind_to(self, target_indent, copy_from : CommonToken):
        self.dents_buffer.append(self.create_token(self.nl_token, copy_from))
        while True:
            prev_indent = self.indentations.pop(0)
            if prev_indent == target_indent:
                break
            if target_indent > prev_indent:
                self.indentations.insert(0, prev_indent)
                self.dents_buffer.append(self.create_token(self.indent_token, copy_from))
                break
            self.dents_buffer.append(self.create_token(self.dedent_token, copy_from))
        self.indentations.insert(0, target_indent)
        return self.dents_buffer.pop(0)

    def apply(self, t: CommonToken):
        if self.should_ignore_eof:
            self.reached_eof = True
            return t
        else:
            # when we reach EOF, unwind all indentations. If there aren't any, insert a NL. This lets the grammar treat
            # un-indented expressions as just being NL-terminated, rather than NL|EOF.
            if not self.indentations:
                r = self.create_token(self.nl_token, t)
                self.dents_buffer.append(t)
            else:
                r = self.unwind_to(0, t)
                self.dents_buffer.append(t)
            self.reached_eof = True
            return r
